### PR TITLE
feat: LEAP-1198: Add labeling unsaved changes warning

### PR DESCRIPTION
**Pull request by Gondragos**
This introduces a functionality of blocking any navigation while there are some unsaved changes in the labeling config.
image
It provides:
- a new component [`LeaveBlocker`](https://github.com/HumanSignal/label-studio/pull/6100/files#diff-ee0a05137865b6987fccf651917989d4ea300c246aa5c05681cbe5ba010530c4) that manages blocking of all cases of navigation
- a new component [`UnsavedChanges`](https://github.com/HumanSignal/label-studio/pull/6100/files#diff-9435ca481c65c6d489cf3fc5438d121b3854cae0fe51e3eff58961b77ee1d695) that  uses [`LeaveBlocker`](https://github.com/HumanSignal/label-studio/pull/6100/files#diff-ee0a05137865b6987fccf651917989d4ea300c246aa5c05681cbe5ba010530c4) and provides a modal [`unsavedChangesModal`](https://github.com/HumanSignal/label-studio/pull/6100/files#diff-9435ca481c65c6d489cf3fc5438d121b3854cae0fe51e3eff58961b77ee1d695R18-R78) for action confirmation dialog
- [`browserHistory`](https://github.com/HumanSignal/label-studio/pull/6100/files#diff-36c1030ca59fcd90ceb45a672287abc0f8a01670167dc4d269eb7f44425b85b0R29-R44) improvement to handle history blocks from LeaveBlocker and indicate right that it is blocking
- fix for handeling `popstate` at [AsyncPage.js](https://github.com/HumanSignal/label-studio/pull/6100/files#diff-280b7a6559377f7fbf54c5ddc9a3ede887c00f1385bcf71a08e794cd38fa6958) for preventing false positive triggering while history is blocked 
- add unsaved changes [indication](https://github.com/HumanSignal/label-studio/pull/6100/files#diff-b8cade5f083b7ee986ecfffd94abcf8500ea563ec194ccf3ca3d94f2edda8a89L461-R465)
- make [saveConfig](https://github.com/HumanSignal/label-studio/pull/6100/files#diff-2e8f9598df92383916945086751cc4b39267ab0b91c6388d8d30057d5884327aL13-R49) work with ProjectProvider to avoid caching problems
- add useful field into [`ApiProvider`](https://github.com/HumanSignal/label-studio/pull/6100/files#diff-61041d4054a0d499fe672745381b02935f9ca798c8ad74b9b101b480be70c63e) including [it's types](https://github.com/HumanSignal/label-studio/pull/6100/files#diff-2c8e58afc9abf073b453294a64724ac165a6ac65fb15b00a226963d3eed00850)
- allow [`updateProject`](https://github.com/HumanSignal/label-studio/pull/6100/files#diff-546c1a320dbf1e8f7e340a651b5eb959651f010c121a2afb996645206f2e8efcL58-R69) to return errors to handle them in editor
- [fix project caching](https://github.com/HumanSignal/label-studio/pull/6100/files#diff-546c1a320dbf1e8f7e340a651b5eb959651f010c121a2afb996645206f2e8efcL66-R83) on calling updateProject

### PR fulfills these requirements
- [x] Commit message(s) and PR title follows the format `[fix|feat|ci|chore|doc]: TICKET-ID: Short description of change made` ex. `fix: DEV-XXXX: Removed inconsistent code usage causing intermittent errors`
- [ ] Tests for the changes have been added/updated (for bug fixes/features)
- [ ] Docs have been added/updated (for bug fixes/features)
- [x] Best efforts were made to ensure docs/code are concise and coherent (checked for spelling/grammatical errors, commented out code, debug logs etc.)
- [x] Self-reviewed and ran all changes on a local instance (for bug fixes/features)

#### Change has impacts in these area(s)
- [ ] Product design
- [ ] Backend (Database)
- [ ] Backend (API)
- [x] Frontend

#### Does this change affect security?
I do not think so.

#### What feature flags were used to cover this change?
`fflag_feat_front_leap_1198_unsaved_changes_180724`

### Does this PR introduce a breaking change?
_(check only one)_
- [ ] Yes, and covered entirely by feature flag(s)
- [ ] Yes, and covered partially by feature flag(s)
- [ ] No
- [x] Not sure (briefly explain the situation below)

I've tried to avoid breaking changes. But we will see...

### What level of testing was included in the change?
_(check all that apply)_
- [ ] e2e
- [ ] integration
- [ ] unit

### Which logical domain(s) does this change affect?
`Labeling Config`, `ApiProvider`, `ProjectProvider`, `BrowserHistory`

### CODE REVIEW

**Review by yyassi-heartex — APPROVED** [review]
looks good overall - my only request would be to add unit tests for the new component

File: label_studio/feature_flags.json
Comment (hlomzik) [inline label_studio/feature_flags.json:1]: why did you remove this file? it's autogenerated and used on backend as I know

File: web/apps/labelstudio/src/app/App.jsx [hunk 1]
```diff
@@ -17,18 +18,30 @@ import "./App.styl";
...
-import { FF_OPTIC_2, isFF } from "../utils/feature-flags";
+import { FF_OPTIC_2, FF_UNSAVED_CHANGES, isFF } from "../utils/feature-flags";
 import { ToastProvider, ToastViewport } from "../components/Toast/Toast";
 
 const baseURL = new URL(APP_SETTINGS.hostname || location.origin);
+export const UNBLOCK_HISTORY_MESSAGE = "UNBLOCK_HISTORY";
 
 const browserHistory = createBrowserHistory({
   basename: baseURL.pathname || "/",
   getUserConfirmation: (message, callback) => {
```
Comment (hlomzik) [inline web/apps/labelstudio/src/app/App.jsx:30]: what is the purpose of `callback`? is it used to unblock navigation?

File: web/apps/labelstudio/src/app/App.jsx [hunk 2]
```diff
@@ -17,18 +18,30 @@ import "./App.styl";
...
+    const callbackWrapper = (result) => {
+      browserHistory.isBlocking = false;
+      callback(result);
+      isFF(FF_UNSAVED_CHANGES) && window.postMessage({ source: "label-studio", payload: UNBLOCK_HISTORY_MESSAGE });
+    };
     if (isFF(FF_OPTIC_2) && message === DRAFT_GUARD_KEY) {
-      draftGuardCallback.current = callback;
+      draftGuardCallback.current = callbackWrapper;
+    } else if (isFF(FF_UNSAVED_CHANGES) && message === LEAVE_BLOCKER_KEY) {
+      leaveBlockerCallback.current = callbackWrapper;
```
Comment (hlomzik) [inline web/apps/labelstudio/src/app/App.jsx:43]: can we somehow use only one logic for both cases? it's getting messy...

File: label_studio/feature_flags.json
Comment (amymrou) [reply label_studio/feature_flags.json:1]: Looks like it wasn't deleted, it appears it was minified. @Gondragos Let's revert this.

File: web/apps/labelstudio/src/app/App.jsx (hunk 2 above)
Comment (bmartel) [reply web/apps/labelstudio/src/app/App.jsx:43]: This would require refactoring the mechanisms used for blocking/restoration of the DraftGuard, to use the LeaveBlocker instead. Seems like that would be a great option, not sure the amount of changes required would be something tenable inside of the same PR.

File: web/apps/labelstudio/src/app/App.jsx (hunk 1 above)
Comment (Gondragos) [reply web/apps/labelstudio/src/app/App.jsx:30]: Yep. And by calling it with false/true we decide should we prevent going to another "page" or not.

Comment (hlomzik) [reply web/apps/labelstudio/src/app/App.jsx:30]: so can you rename it to reflect semantics? or at least add a comment please

**Review by bmartel — APPROVED** [review]

**Comment by Gondragos** [comment]
/git merge develop
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/10165081523)
> Successfully merged:  17 files changed, 28 insertions(+), 632 deletions(-)

**Review by bmartel — APPROVED** [review]

**Review by MihajloHoma — APPROVED** [review]
